openedx/core/djangoapps/content_libraries/tasks.py

```diff
@@ -169,6 +174,70 @@ def duplicate_children(
                 self.status.fail({'raw_error_msg': str(exception)})
 
 
+@shared_task(base=LoggedTask)
+@set_code_owner_attribute
+def create_or_update_xblock_upstream_link(usage_key):
+    """
+    Create or update upstream link for a single xblock.
+    """
+    ensure_cms("create_or_update_xblock_upstream_link may only be executed in a CMS context")
+    xblock = modulestore().get_item(UsageKey.from_string(usage_key))
+    if not xblock.upstream or not xblock.upstream_version:
+        TASK_LOGGER.info(f"No upstream or upstream_version found for xblock: {xblock.usage_key}")
+        return
+    try:
+        course_name = CourseOverview.get_from_id(xblock.course_id).display_name_with_default
+    except CourseOverview.DoesNotExist:
+        TASK_LOGGER.exception(f'Could not find course: {xblock.course_id}')
+        return
+    api.create_or_update_xblock_upstream_link(xblock, str(xblock.course_id), course_name)
+
+
+@shared_task(base=LoggedTask)
+@set_code_owner_attribute
+def create_or_update_upstream_links(course_key_str: str, force: bool = False, created: datetime | None = None):
+    """
+    A Celery task to create or update upstream downstream links in database from course xblock content.
+    """
+    ensure_cms("create_or_update_upstream_links may only be executed in a CMS context")
+
+    if not created:
+        created = datetime.now(timezone.utc)
+    course_status = get_or_create_course_link_status(course_key_str, created)
+    if course_status.status in [
+        CourseLinksStatusChoices.COMPLETED,
+        CourseLinksStatusChoices.PROCESSING
+    ] and not force:
+        return
+    store = modulestore()
+    course_key = CourseKey.from_string(course_key_str)
+    course_status.status = CourseLinksStatusChoices.PROCESSING
+    course_status.save()
+    try:
+        course_name = CourseOverview.get_from_id(course_key).display_name_with_default
+    except CourseOverview.DoesNotExist:
+        TASK_LOGGER.exception(f'Could not find course: {course_key_str}')
```

Comment:
ditto nit: do we need to log the whole exception? I don't think it will tell us much more than the error message does.
And shouldn't we mark the task as failed?

```suggestion
        TASK_LOGGER.error(f'Could not find course: {course_key_str}')
        course_status.status = CourseLinksStatusChoices.FAILED
        course_status.save()
```

Reply:
Same here.

Reply:
How about marking the `course_status` as failed if it fails?

Reply:
Nice catch! Updated.